src/shared/sam/debugger/goSamDebug.ts

```diff
@@ -0,0 +1,185 @@
+/*!
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import * as os from 'os'
+import * as path from 'path'
+import * as vscode from 'vscode'
+import { GoDebugConfiguration, GO_DEBUGGER_PATH, isImageLambdaConfig } from '../../../lambda/local/debugConfiguration'
+import { RuntimeFamily } from '../../../lambda/models/samLambdaRuntime'
+import * as pathutil from '../../../shared/utilities/pathUtils'
+import { ExtContext } from '../../extensions'
+import { findParentProjectFile } from '../../utilities/workspaceUtils'
+import { DefaultSamLocalInvokeCommand, WAIT_FOR_DEBUGGER_MESSAGES } from '../cli/samCliLocalInvoke'
+import { invokeLambdaFunction, makeInputTemplate, waitForPort } from '../localLambdaRunner'
+import { SamLaunchRequestArgs } from './awsSamDebugger'
+import { getLogger } from '../../logger'
+import { chmod, ensureDir, writeFile } from 'fs-extra'
+import { ChildProcess } from '../../utilities/childProcess'
+import { Timeout } from '../../utilities/timeoutUtils'
+import { SystemUtilities } from '../../../shared/systemUtilities'
+
+/**
+ * Launches and attaches debugger to a SAM Go project.
+ */
+export async function invokeGoLambda(ctx: ExtContext, config: GoDebugConfiguration): Promise<GoDebugConfiguration> {
+    config.samLocalInvokeCommand = new DefaultSamLocalInvokeCommand([WAIT_FOR_DEBUGGER_MESSAGES.GO_DELVE])
+    // eslint-disable-next-line @typescript-eslint/unbound-method
+    config.onWillAttachDebugger = goToSleep
+
+    if (!config.noDebug) {
+        installDebugger(config.debuggerPath!)
+    }
+
+    const c = (await invokeLambdaFunction(ctx, config, async () => {})) as GoDebugConfiguration
+    return c
+}
+
+/**
+ * Triggered before the debugger attachment process begins. We should verify that the debugger is ready to go
+ * before returning. We'll wait a little bit before checking ports, just to make sure the debugger doesn't break.
+ *
+ * @param debugPort Port to check for activity
+ * @param timeout Cancellation token to prevent stalling
+ */
+async function goToSleep(debugPort: number, timeout: Timeout) {
```

Comment:
Can you give this a more descriptive name? Also explain why more why this is now needed?